permission_validators/src/lib.rs

```diff
@@ -3,17 +3,30 @@
 #![no_std]
 #![no_main]
 
-extern crate alloc;
 #[cfg(not(test))]
 extern crate panic_halt;
 
 use iroha_validator::prelude::*;
 
-/// Validate any [`Instruction`](iroha_validator::data_model::isi::Instruction).
-//
-// TODO: Not exhaustive list. Add more validators here.
-#[entrypoint(params = "[authority, instruction]")]
-pub fn validate(authority: <Account as Identifiable>::Id, instruction: Instruction) -> Verdict {
+/// Validate operation.
+#[entrypoint(params = "[authority, operation]")]
+pub fn validate(
+    authority: <Account as Identifiable>::Id,
+    operation: NeedsPermissionBox,
+) -> Verdict {
+    validate_recursively(
+        authority,
+        operation,
+        |_authority, _transaction| pass!(),
+        validate_instruction,
+        |_authority, _query| pass!(),
+    )
+}
```

Comment:
The code for validating instructions was badly structured even before you the started the work on dynamic permission validators and as a result there is a lot of redundant computation. The fact that all validators will be invoked for `Instruction::Fail` doesn't sound right to me. The crux of the problem is that `Instruction` (like `QueryBox`) is only a serialization format and should be polymorphically dispatched as soon as possible. I'll show you the beginning of how it should be done so you can figure it out from there:

```suggestion
pub fn validate(
    authority: <Account as Identifiable>::Id,
    instruction: Instruction,
) -> Verdict {
    instruction.validate()
}

pub trait Validate {
    type Error = Error;
    fn validate(self, authority: <Account as Identifiable>::Id) -> Verdict;
}

impl Validate for InstructionBox {
    type Error = Error;

    fn validate(self, authority: <Account as Identifiable>::Id) -> Verdict {
        use InstructionBox::*;

        match self {
            Register(register_box) => register_box.validate(authority, wsv),
            Unregister(unregister_box) => unregister_box.validate(authority, wsv),
            Mint(mint_box) => mint_box.validate(authority, wsv),
            Burn(burn_box) => burn_box.validate(authority, wsv),
            Transfer(transfer_box) => transfer_box.validate(authority, wsv),
            If(if_box) => if_box.validate(authority, wsv),
            Pair(pair_box) => pair_box.validate(authority, wsv),
            Sequence(sequence) => sequence.validate(authority, wsv),
            Fail(fail_box) => fail_box.validate(authority, wsv),
            SetKeyValue(set_key_value) => set_key_value.validate(authority, wsv),
            RemoveKeyValue(remove_key_value) => remove_key_value.validate(authority, wsv),
            Grant(grant_box) => grant_box.validate(authority, wsv),
            Revoke(revoke_box) => revoke_box.validate(authority, wsv),
            ExecuteTrigger(execute_trigger) => execute_trigger.validate(authority, wsv),
            SetParameter(parameter_box) => parameter_box.validate(authority, wsv),
            NewParameter(parameter_box) => parameter_box.validate(authority, wsv),
        }
    }
}
```

For inspiration you can look at how `Execute` trait in `core/smartcontracts/mod.rs` is implemented. This is exact 1-to-1 correspondence to how `Validate` trait should be implemented

Reply:
I don't like that idea.

1. It doesn't introduce anything new. We already can ensure that every instruction is validated just by using `match`.
2. It will be impossible to split such implementations into differenct crates because of orphan rule. I explained why we still need things like `iroha_asset_burn_validator` [here](https://github.com/hyperledger/iroha/pull/3329#discussion_r1155765355)
3. It introduces another semantics of validation. It becomes more like `Instruction` propery. However in fact it's not. It the validator responsibility

Reply:
However we indeed can make things like `iroha_asset_burn_validator` more strongly typed by replacing instruction parameter type with a concrete type

Reply:
Maybe you want to have such `match` inside internal function like `validate_recursively` and to force user to provide validation function for each instruction type?

Reply:
>     1. It doesn't introduce anything new. We already can ensure that every instruction is validated just by using `match`.

it's not that some instruction is not validated, it's that some instructions are unnecessarily validated. In the current implementation every validator function is invoked for every instruction. That should no happen

Reply:
> Maybe you want to have such match inside internal function like validate_recursively and to force user to provide validation function for each instruction type?

no, in my proposal nothing is changed from the perspective of the user

> 2. It will be impossible to split such implementations into differenct crates because of orphan rule. I explained why we still need things like iroha_asset_burn_validator https://github.com/hyperledger/iroha/pull/3329#discussion_r1155765355

I don't see any benefit in one crate per validator approach

> 3. It introduces another semantics of validation. It becomes more like Instruction propery. However in fact it's not. It the validator responsibility

it's just the visitor pattern, very common in OOP

Reply:
> no, in my proposal nothing is changed from the perspective of the user

actually, you were correct here. Yes, there should be one validation function for every instruction

Reply:
Check out new updates